### PR TITLE
feat: add locked amount to position API and add lock info API

### LINKED ISSUE #1286: Additional features for locking and unlocking
## Is your feature request related to a problem? Please describe.

- There is a request to reference data for locking and unlocking account assets.

### 1. Storing data of locked balances
- #1293
- #1300
  - New TBL
    - `locked_position`
    - `lock`
    - `unlock`
  - Fix indexer batch
    - indexer_Position_Bond.py
    - indexer_Position_Share.py

### 2. Add locked balances to the return item of Position API
- #1300
  - Balances held by account address
    - GET: /Position/{account_address}/Share/
    - GET: /Position/{account_address}/StraightBond/
    - GET: /Position/{account_address}/Share/{token_address}
    - GET: /Position/{account_address}/StraightBond/{token_address}
  - List and individual references of token holders
    - GET: /Token/{token_address}/Holders/Collection/{list_id}
- #1293
  - List and individual references of token holders
    - GET: /Token/{token_address}/Holders
  - Change of conditions for calculation of number of holders
    - GET: /Token/{token_address}/Holders/Count

### 3. Add transfer indexing and holders collection creation process to watch for unlock events
- #1300
  - Add transfer history when unlock event occurs
    - indexer_Transfer.py
  - Reflecting the holder change when the lock/unlock event occurs in the data of the holder collection
    - indexer_Token_Holders.py

### 4. List all account's locked positions and events
- #1300
  - [NEW] GET: /Position/{account_address}/Share/Lock
  - [NEW] GET: /Position/{account_address}/StraightBond/Lock
  - [NEW] GET: /Position/{account_address}/Share/Lock/Event
  - [NEW] GET: /Position/{account_address}/StraightBond/Lock/Event

### PR DESCRIPTION
**Pull request by purplesmoke05**
#1286, #1295 

#### API

- contract_log
  - GET: `/Events/IbetSecurityTokenInterface/{token_address}`
    - **New API**
  - GET: `/Events/E2EMessaging`
    - Added search block number range upper limit (10000).
  - GET: `/Events/E2EMessaging`
    - Added search block number range upper limit (10000).
  - GET: `/Events/E2EMessaging`
    - Added search block number range upper limit (10000).
- user_position
  - GET: `/Locked`
    - **New API**
  - GET: `/Position/{account_address}/StraightBond`
    - Add `locked` attribute in response schema.
  - GET: `/Position/{account_address}/StraightBond/{token_address}`
    - Add `locked` attribute in response schema.
  - GET: `/Position/{account_address}/Share`
    - Add `locked` attribute in response schema.
  - GET:  `/Position/{account_address}/Share/{token_address}`
    - Add `locked` attribute in response schema.
- token_info
  - GET: `/Token/{token_address}/Holders`
    - Add `locked` attribute in response schema.
  - GET: `/Token/{token_address}/Holders/Count`
    - Include locked balances in holder count

#### Batch

- `indexer_Position_Bond.py`
`indexer_Position_Share.py`
  - Modified to sync `locked` amount.
- `indexer_Token_Holders.py` 
  - Modified to track `unlock` event.

#### DB

- TBL: `locked_position`
  - Newly added.

### CODE REVIEW

File: app/model/db/idx_locked.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+class IDXLocked(Base):
+    """Token Locked Amount (INDEX)"""
+    __tablename__ = "locked"
+
+    # Token Address
+    token_address = Column(String(42), primary_key=True)
+    # Lock Address
+    lock_address = Column(String(42), primary_key=True)
+    # Account Address
+    account_address = Column(String(42), primary_key=True)
```
Comment (purplesmoke05) [inline app/model/db/idx_locked.py:37]: PK is set to `token_address`, `lock_address` and `account_address`.

File: batch/indexer_Position_Bond.py [hunk 2]
```diff
@@ -343,6 +344,27 @@ def __sync_lock(self, db_session: Session, block_to: int):
...
+                for lock_address in lock_map:
+                    for account_address in lock_map[lock_address]:
+                        value = self.__get_account_locked_token(token, lock_address, account_address)
+                        self.__sink_on_locked(
+                            db_session=db_session,
+                            token_address=to_checksum_address(token.address),
+                            lock_address=lock_address,
+                            account_address=account_address,
+                            value=value
+                        )
```
Comment (purplesmoke05) [inline batch/indexer_Position_Bond.py:365]: Retrieving account address and lock address from events to fetch locked balance at the point of the batch cycle.

File: app/model/schema/position.py [hunk 3]
```diff
@@ -57,6 +62,7 @@ class SecurityTokenPosition(BaseModel):
     pending_transfer: int
     exchange_balance: int
     exchange_commitment: int
+    locked: Optional[int] = Field(default=None, description="set when enable_index=true")
```
Comment (purplesmoke05) [inline app/model/schema/position.py:65]: `locked`value in Position API is set only when `enable_index` is `true`.

File: app/api/routers/token.py [hunk 4]
```diff
@@ -549,3 +557,164 @@ def list_all_transfer_approval_histories(
...
 
+
+@router.get(
+    "/{token_address}/Lock",
+    summary="Token Lock Events",
+    operation_id="Lock",
+    response_model=GenericSuccessResponse[LockEventsResponse],
+    responses=get_routers_responses(DataNotExistsError, InvalidParameterError)
+)
+def list_all_lock_events(
```
Comment (purplesmoke05) [inline app/api/routers/token.py:568]: Add event list endpoints for lock and unlock.

File: app/api/routers/events.py [hunk 5]
```diff
@@ -57,12 +66,18 @@
...
+        RequestBlockRangeLimitExceededError
+    )
 )
 def list_all_e2e_messaging_event_logs(
     request_query: E2EMessagingEventsQuery = Depends()
 ):
     """List all E2EMessaging event logs"""
+    if request_query.to_block - request_query.from_block > REQUEST_BLOCK_RANGE_LIMIT:
+        raise RequestBlockRangeLimitExceededError("Search request range is over the limit")
+
```
Comment (purplesmoke05) [inline app/api/routers/events.py:80]: Related to #1295.
Added upper limit for event query search range.

File: app/api/routers/events.py [hunk 6]
```diff
@@ -272,3 +299,81 @@ def list_all_ibet_security_token_escrow_event_logs(
...
+    "/IbetSecurityTokenInterface/{token_address}",
+    summary="List all IbetSecurityTokenInterface event logs",
+    operation_id="IbetSecurityTokenEscrowEvents",
+    response_model=GenericSuccessResponse[ListAllEventsResponse],
+    responses=get_routers_responses(
+        InvalidParameterError,
+        RequestBlockRangeLimitExceededError
+    )
+)
+def list_all_ibet_security_token_interface_event_logs(
```
Comment (purplesmoke05) [inline app/api/routers/events.py:315]: Added API for `IbetSecurityTokenInterface`.
You can use this to fetch lock/unlock events about given `token_address`.

File: app/contracts/abi.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+ABIDescription = Union[ABIEventDescription, ABIFunctionDescription, ABIGenericDescription]
+ABI = List[ABIDescription]
+
+
+@lru_cache(None)
+def create_abi_event_argument_models(contract_name: str) -> list[Type[BaseModel]]:
+    contract_file = f"app/contracts/json/{contract_name}.json"
+
+    contract_json = json.load(open(contract_file, 'r'))
+    abi_list = parse_obj_as(ABI, contract_json.get("abi"))
```
Comment (purplesmoke05) [inline app/contracts/abi.py:130]: Added a function to generate parser for ABI.
Event arguments filter models are generated in this function and used to validate in `/Events` API logic.

File: app/api/routers/token.py (hunk 4 above)
Comment (purplesmoke05) [reply app/api/routers/token.py]: GET: `/Token/{token_address}/Lock`/ `/Token/{token_address}/Unlock` are removed:  https://github.com/BoostryJP/ibet-Wallet-API/pull/1293/commits/5cc1a101083132bbba5632db112c61b48ac79443

**Comment by YoshihitoAso** [comment]
https://github.com/BoostryJP/ibet-Wallet-API/pull/1293/commits/9ff237243e5cbf92bea8b9f78b971e349bde8cac
Additional implementation and a bit of refactoring.

**Review by YoshihitoAso — APPROVED** [review]

**Pull request merged**